### DISCUSSION TITLE
creating coco formatted result.json with sahi predict CLI command for coco evaluation

### DISCUSSION THREAD

**Discussion by isspid** [obss/sahi#336] [Q&A] [answered]
I thought a question rather than an issue was more fitting for this.

I am a little bit confused by the CLI commands. I have a model which I use to run inference on a directory of images. After running it I can get the new images which have the predictions visualized. Could you please explain how to get a .json file of these predictions so that I can run the coco_evaluate command afterwards?

Edit: Basically I want to know the procedure of getting the results.json file.

**Comment by fcakyon** [comment, marked as answer]
@isspid thanks for bringing this up. This has been answered before: https://github.com/obss/sahi/issues/273#issuecomment-971402591

But I will briefly explain again 🚀 

1. You have to prepare a COCO formatted dataset (image folder and COCO formatted `dataset.json`).
2. ```sahi predict --source coco/image/folder --dataset_json_path coco_dataset.json + other arguments``` will export a `results.json`.
3. using `dataset.json` and `result.json` you run ```sahi coco evaluate --dataset_json_path dataset.json --result_json_path result.json```.
4. `eval.json` containing coco evaluation results will be exported.
5. (bonus) you can generate coco error analysis plots by ```sahi coco analyse --dataset_json_path dataset.json --result_json_path result.json```.

**Reply by isspid** [reply]
I will try it now and let you know how it goes.

Is there a way we can support your work financially? I can't afford much as I am just a student but it feels dirty to get this much help for free haha.

**Reply by fcakyon** [reply]
I will be preparing gif documentation for these steps this week 🔥

**Reply by fcakyon** [reply]
@isspid all is free of charge 🚀 best thing you can do in return is:
✅  star the repo
✅  share the repo url in your social media to increase the reach
✅ tell your friends and lecturers about this repo

**Reply by isspid** [reply]
So I ran the commands and they went smooth but I have the problem that my eval.json file looks like this:
![image](https://user-images.githubusercontent.com/91065022/148390251-afefeb20-ff5c-4b35-8bf9-1f93c845db32.png)
even though I am sure that the predictions are much much better.

For example, this is a picture from the folder called visuals_with_gt:
![image](https://user-images.githubusercontent.com/91065022/148390480-7ad62b8a-5693-45bb-ab29-92e70c58d708.png)

I suspect that the problem is with my dataset.json file, which I got from using CVAT, the tool I use for labelling. 
[dataset.zip](https://github.com/obss/sahi/files/7822119/dataset.zip) <-- this is my dataset.json file

**Reply by fcakyon** [reply]
@isspid yes the issue is in your dataset. Your category ids start from 1 while model is performing inference on starting with category 0.

You can update category ids as:

```python
from sahi.utils.coco import Coco
from sahi.utils.file import save_json

# init Coco objects by specifying coco dataset paths and image folder directories
coco = Coco.from_coco_dict_or_path("dataset.json")

# select window categories; and map them to id 0
desired_name2id = {
  "window": 0,
}
coco.update_categories(desired_name2id)

# export updated/filtered COCO dataset
save_json(coco.json, "updated_dataset.json")
```

**Reply by isspid** [reply]
Fatih, you have been amazing once again. Many thanks.

**Reply by dimpolitik** [reply]
@isspid I have two questions. Which format do you use in CVAT to create the dataset.json and then which snippet code to produce the result.json file?
In my case I saved the annotations in COCO fromat and then I used this to produce the result.json.
 
sahi predict --slice_width 640 --slice_height 640 --overlap_height_ratio 0.2 --overlap_width_ratio 0.2 --model_confidence_threshold 0.5 --source /home/politikos/datasets/pets/images/test/ --dataset_json_path dataset_coco.json --model_path /home/politikos/pets_detection/runs/detect/pets_xl/weights/best.pt --model_type yolov8 --model_config_path /home/politikos/.local/lib/python3.11/site-packages/ultralytics/cfg/datasets/pets.yaml. 

But the dataset_coco.json and the resulting result.json file don't have the same "content format" when I open them. 
Thank you for your time!

**Reply by isspid** [reply]
Hi @dimpolitik. It has been quite some time but I will try to help as much as I can. 

From CVAT you can export the COCO formatted json file by selecting the setting below:
![grafik](https://github.com/obss/sahi/assets/91065022/66ed1414-929a-4b7a-af00-8f5cdd5d210f)

In order to generate the results.json file please run `sahi predict --source coco/image/folder --dataset_json_path coco_dataset.json + other arguments`

If they don't have the same format, please upload them here, I will try to take a look at them and help in anyway I can.

**Reply by dimpolitik** [reply]
Hi @isspid! 
Attached the two files.
The result json was produced with: 

sahi predict --source /home/politikos/datasets/pets/images/test/ --dataset_json_path dataset.json --model_path /home/politikos/pets_detection/runs/detect/pets_x16/weights/best.pt --model_type yolov8 --model_config_path /home/politikos/.local/lib/python3.11/site-packages/ultralytics/cfg/datasets/pets.yaml
[dataset.json](https://github.com/obss/sahi/files/13217936/dataset.json)
[result.json](https://github.com/obss/sahi/files/13217937/result.json)


When I run:  sahi coco evaluate --type bbox --classwise --iou_thrs 0.5 --dataset_json_path dataset.json --result_json_path result.json
I get zero values and -1.0 for APs.

Important notice. I trained the YOLO model with the yolo txt annotation files, not coco if this matters.

Thank you so much for your willingness to help. Really appreciated!

**Reply by isspid** [reply]
I believe you are experiencing the same problem as me, which is discussed in this thread. Please take a look at this answer:
https://github.com/obss/sahi/discussions/336#discussioncomment-1918853.

I will simply explain here what @fcakyon explained to me in the answer.

Basically "your category ids start from 1 while model is performing inference on starting with category 0."

You can update the category ids:

```python
from sahi.utils.coco import Coco
from sahi.utils.file import save_json

# init Coco objects by specifying coco dataset paths and image folder directories
coco = Coco.from_coco_dict_or_path("dataset.json")

# select window categories; and map them to id 0
desired_name2id = {
  "window": 0,
}
coco.update_categories(desired_name2id)

# export updated/filtered COCO dataset
save_json(coco.json, "updated_dataset.json")
```

**Comment by dimpolitik** [comment]
I was aware of this code. The corrected json is the attached. I get the same -1 values.
[correct_dataset.json](https://github.com/obss/sahi/files/13218316/correct_dataset.json)

Did your corrected json have the same structure?

**Comment by isspid** [comment]
Your corrected json doesn't have any annotations as far as I can see.

I think you need to edit the code in order for it to work for you. Please try this and let me know how it goes.

```python
from sahi.utils.coco import Coco
from sahi.utils.file import save_json

# init Coco objects by specifying coco dataset paths and image folder directories
coco = Coco.from_coco_dict_or_path("dataset.json")

# select window categories; and map them to id 0
desired_name2id = {
  "shark": 0,
  "skate": 1,
}
coco.update_categories(desired_name2id)

# export updated/filtered COCO dataset
save_json(coco.json, "updated_dataset.json")
```

**Reply by dimpolitik** [reply]
That's true!!! I thought that "window" : 0 was the default way to correct json. Now I get values other than -1. Great!! Thank you so much! I was struggling many days with that.

"bbox_mAP":0.242,
    "bbox_mAP50":0.479,
    "bbox_mAP75":0.209,
    "bbox_mAP_s":-1.0,
    "bbox_mAP_m":-1.0,
    "bbox_mAP_l":0.242,
    "bbox_mAP50_s":-1.0,
    "bbox_mAP50_m":-1.0,
    "bbox_mAP50_l":0.479,
    "bbox_mAP_copypaste":"0.242 0.209 0.479 -1.000 -1.000 0.242 -1.000 -1.000 0.479",
    "results_per_category":{
        "bbox_shark_mAP":"0.174",
        "bbox_shark_mAP_s":"-1.000",
        "bbox_shark_mAP_m":"-1.000",
        "bbox_shark_mAP_l":"0.174",
        "bbox_shark_mAP50":"0.354",
        "bbox_shark_mAP50_s":"-1.000",
        "bbox_shark_mAP50_m":"-1.000",
        "bbox_shark_mAP50_l":"0.354",
        "bbox_skate_mAP":"0.310",
        "bbox_skate_mAP_s":"-1.000",
        "bbox_skate_mAP_m":"-1.000",
        "bbox_skate_mAP_l":"0.310",
        "bbox_skate_mAP50":"0.604",
        "bbox_skate_mAP50_s":"-1.000",
        "bbox_skate_mAP50_m":"-1.000",

**Reply by isspid** [reply]
Happy to read it worked. Good luck.

**Reply by bshakya77** [reply]
Hello @isspid,

I used provided code snippet to remap coco128 dataset to id:0. It got converted, but the annotation list is empty as you mentioned above. The code snippet I used is:

from sahi.utils.coco import Coco
from sahi.utils.file import save_json

coco = Coco.from_coco_dict_or_path("coco_instances_val2017.json")

desired_name2id = {
    "person": 0,
    "bicycle": 1,
    "car": 2,
    "motorcycle": 3,
    "airplane": 4,
    "bus": 5,
    "train": 6,
    "truck": 7,
    "boat": 8,
    "traffic light": 9,
    "fire hydrant": 10,
    "stop sign": 11,
    "parking meter": 12,
    "bench": 13,
    "bird": 14,
    "cat": 15,
    "dog": 16,
    "horse": 17,
    "sheep": 18,
    "cow": 19,
    "elephant": 20,
    "bear": 21,
    "zebra": 22,
    "giraffe": 23,
    "backpack": 24,
    "umbrella": 25,
    "handbag": 26,
    "tie": 27,
    "suitcase": 28,
    "frisbee": 29,
    "skis": 30,
    "snowboard": 31,
    "sports ball": 32,
    "kite": 33,
    "baseball bat": 34,
    "baseball glove": 35,
    "skateboard": 36,
    "surfboard": 37,
    "tennis racket": 38,
    "bottle": 39,
    "wine glass": 40,
    "cup": 41,
    "fork": 42,
    "knife": 43,
    "spoon": 44,
    "bowl": 45,
    "banana": 46,
    "apple": 47,
    "sandwich": 48,
    "orange": 49,
    "broccoli": 50,
    "carrot": 51,
    "hot dog": 52,
    "pizza": 53,
    "donut": 54,
    "cake": 55,
    "chair": 56,
    "couch": 57,
    "potted plant": 58,
    "bed": 59,
    "dining table": 60,
    "toilet": 61,
    "tv": 62,
    "laptop": 63,
    "mouse": 64,
    "remote": 65,
    "keyboard": 66,
    "cell phone": 67,
    "book": 68,
    "clock": 69,
    "vase": 70,
}
coco.update_categories(desired_name2id)
save_json(coco.json, "updated_dataset.json")

The original dataset file is
[coco_instances_val2017.json](https://github.com/user-attachments/files/18390204/coco_instances_val2017.json)

The updated dataset file I generated with snippet is as follows:
[updated_dataset.json](https://github.com/user-attachments/files/18390189/updated_dataset.json)

I would be grateful if you could suggest what was the issue of generating the empty annotation list.

Regards,
Bijay

**Reply by fcakyon** [reply]
@bshakya77 your category names are not "person", "bicycle" etc. They are named as "0" "1" etc. 

Fix your category names in json or fix your `desired_name2id`.

from your coco_instances_val2017,json:
```json
"categories": [{"id": 1, "name": "0", "supercategory": "0"}, {"id": 2, "name": "1", "supercategory": "1"}, {"id": 3, "name": "2", "supercategory": "2"}, {"id": 4, "name": "3", "supercategory": "3"}, {"id": 5, "name": "4", "supercategory": "4"}, {"id": 6, "name": "5", "supercategory": "5"}, {"id": 7, "name": "6", "supercategory": "6"}, {"id": 8, "name": "7", "supercategory": "7"}, {"id": 9, "name": "8", "supercategory": "8"}, {"id": 10, "name": "9", "supercategory": "9"}, {"id": 11, "name": "10", "supercategory": "10"}, {"id": 12, "name": "11", "supercategory": "11"}, {"id": 13, "name": "12", "supercategory": "12"}, {"id": 14, "name": "13", "supercategory": "13"}, {"id": 15, "name": "14", "supercategory": "14"}, {"id": 16, "name": "15", "supercategory": "15"}, {"id": 17, "name": "16", "supercategory": "16"}, {"id": 18, "name": "17", "supercategory": "17"}, {"id": 19, "name": "18", "supercategory": "18"}, {"id": 20, "name": "19", "supercategory": "19"}, {"id": 21, "name": "20", "supercategory": "20"}, {"id": 22, "name": "21", "supercategory": "21"}, {"id": 23, "name": "22", "supercategory": "22"}, {"id": 24, "name": "23", "supercategory": "23"}, {"id": 25, "name": "24", "supercategory": "24"}, {"id": 26, "name": "25", "supercategory": "25"}, {"id": 27, "name": "26", "supercategory": "26"}, {"id": 28, "name": "27", "supercategory": "27"}, {"id": 29, "name": "28", "supercategory": "28"}, {"id": 30, "name": "29", "supercategory": "29"}, {"id": 31, "name": "30", "supercategory": "30"}, {"id": 32, "name": "31", "supercategory": "31"}, {"id": 33, "name": "32", "supercategory": "32"}, {"id": 34, "name": "33", "supercategory": "33"}, {"id": 35, "name": "34", "supercategory": "34"}, {"id": 36, "name": "35", "supercategory": "35"}, {"id": 37, "name": "36", "supercategory": "36"}, {"id": 38, "name": "37", "supercategory": "37"}, {"id": 39, "name": "38", "supercategory": "38"}, {"id": 40, "name": "39", "supercategory": "39"}, {"id": 41, "name": "40", "supercategory": "40"}, {"id": 42, "name": "41", "supercategory": "41"}, {"id": 43, "name": "42", "supercategory": "42"}, {"id": 44, "name": "43", "supercategory": "43"}, {"id": 45, "name": "44", "supercategory": "44"}, {"id": 46, "name": "45", "supercategory": "45"}, {"id": 47, "name": "46", "supercategory": "46"}, {"id": 48, "name": "47", "supercategory": "47"}, {"id": 49, "name": "48", "supercategory": "48"}, {"id": 50, "name": "49", "supercategory": "49"}, {"id": 51, "name": "50", "supercategory": "50"}, {"id": 52, "name": "51", "supercategory": "51"}, {"id": 53, "name": "52", "supercategory": "52"}, {"id": 54, "name": "53", "supercategory": "53"}, {"id": 55, "name": "54", "supercategory": "54"}, {"id": 56, "name": "55", "supercategory": "55"}, {"id": 57, "name": "56", "supercategory": "56"}, {"id": 58, "name": "57", "supercategory": "57"}, {"id": 59, "name": "58", "supercategory": "58"}, {"id": 60, "name": "59", "supercategory": "59"}, 
```

**Reply by bshakya77** [reply]
Hello @fcakyon and @isspid,

1. I downloaded the pretrained YOLO11n model from the official ultralytics repo.

https://github.com/ultralytics/assets/releases/download/v8.3.0/yolo11n.pt


2. I used the subset of coco dataset (30 data), which has categories name as "person", "bicycle" etc. When I used the whole COCO dataset, it was giving `out of index` error. When I used the following json files as datasets and result file I generated after prediction. The evaluation result came out to be all 0. The visualization with ground truth is coming right.

[instances_val2017_coco30.json](https://github.com/user-attachments/files/18405327/instances_val2017_coco30.json)
[result_coco30.json](https://github.com/user-attachments/files/18405350/result_coco30.json)

![image](https://github.com/user-attachments/assets/2a0fb0c2-3ae8-4b89-9556-4daea5c052ef)

In this process, I see that my instances_val2017_coco30.json starts from category_id: 1. So, I used the following snippet

`
from sahi.utils.coco import Coco
from sahi.utils.file import save_json

coco = Coco.from_coco_dict_or_path("../data/coco2017_subset30/annotations/instances_val2017.json")
desired_name2id = {
    "person": 0,
    "bicycle": 1,
    "car": 2,
    "motorcycle": 3,
    "airplane": 4,
    "bus": 5,
    "train": 6,
    "truck": 7,
    "boat": 8,
    "traffic light": 9,
    "fire hydrant": 10,
    "stop sign": 11,
    "parking meter": 12,
    "bench": 13,
    "bird": 14,
    "cat": 15,
    "dog": 16,
    "horse": 17,
    "sheep": 18,
    "cow": 19,
    "elephant": 20,
    "bear": 21,
    "zebra": 22,
    "giraffe": 23,
    "backpack": 24,
    "umbrella": 25,
    "handbag": 26,
    "tie": 27,
    "suitcase": 28,
    "frisbee": 29,
    "skis": 30,
    "snowboard": 31,
    "sports ball": 32,
    "kite": 33,
    "baseball bat": 34,
    "baseball glove": 35,
    "skateboard": 36,
    "surfboard": 37,
    "tennis racket": 38,
    "bottle": 39,
    "wine glass": 40,
    "cup": 41,
    "fork": 42,
    "knife": 43,
    "spoon": 44,
    "bowl": 45,
    "banana": 46,
    "apple": 47,
    "sandwich": 48,
    "orange": 49,
    "broccoli": 50,
    "carrot": 51,
    "hot dog": 52,
    "pizza": 53,
    "donut": 54,
    "cake": 55,
    "chair": 56,
    "couch": 57,
    "potted plant": 58,
    "bed": 59,
    "dining table": 60,
    "toilet": 61,
    "tv": 62,
    "laptop": 63,
    "mouse": 64,
    "remote": 65,
    "keyboard": 66,
    "cell phone": 67,
    "microwave": 68,
    "oven": 69,
    "toaster": 70,
    "sink": 71,
    "refrigerator": 72,
    "book": 73,
    "clock": 74,
    "vase": 75,
    "scissors": 76,
    "teddy bear": 77,
    "hair drier": 78,
    "toothbrush": 79
}
coco.update_categories(desired_name2id)
save_json(coco.json, "updated_dataset_coco30.json")
`
The following is the updated json dataset file that I generated:
[updated_dataset_coco30.json](https://github.com/user-attachments/files/18405384/updated_dataset_coco30.json)

It does convert the categories to id: 0. However, when I used this updated json file for evaluation, I get the following error.

`AssertionError: Results do not correspond to current coco set`

My question:

Is the evaluation result came out as 0  because I used the pretrained model trained on full COCO dataset and I used instances_val2017_coco30 for prediction and evaluation. Does it matters?

Or, is it still because the mismatch in categories in dataset.json and result.json file that I used.

Could you please provide me with the code to convert the original COCO dataset (with category_id: 1) to COCO dataset starting category id: 0?

I highly appreciate your feedback.

Regards,
Bijay
